### ISSUE TITLE
Investigate story regarding configurable event logging

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Starcounter#3064]
We have a need for a general way to define that certain *actions* happening in a particular application should generate *logs*. The exact nature of what be in every such log is context-bound.

As a simple example, we need to support systems that need to contain logic such as "when an artifact with state x change to state y, we want a log to be generated including the timestamp and the current user".

Our idea is to make this a general purpose application.

I've started outlining the needs and have some initial ideas on how this should be implemented.

**Comment by eriohl** [comment]
Sounds like a (the) transaction log (although no user in this since we lack a built concept of users.).

**Comment by per-samuelsson** [comment]
> although no user in this since we lack a built concept of users.

Yeah, but the idea is not to support any build in user, but to support a context defined by the application developer. So to support this requirement, we need to support running some kind of user defined code on such event.

Some kind of transaction log subscription to drive it might be the best way to implement it. The idea I had build on code-generated callbacks + commit hooks. I'm not all through it yet, so I don't know if it will turn out good or bad. We can have a discussion about it when I've given in some more thought, and then compare alternatives.

**Comment by careri** [comment]
What's the best way of achieving this at the moment? Are there commit hooks?
@per-samuelsson

**Comment by careri** [comment]
I've figured out how to register the Hook<T>.CommitInsert.

But it seems to me like this hook is trigger during the commit? Not after the commit is written. So running a task using the new item on another thread (DbSession.RunAsync) will throw an exception InvalidObjectAccess exception.

Is this correct?
@eriohl  @per-samuelsson

**Comment by eriohl** [comment]
Commit hooks are triggered at commit (not after commit completed) yes.

**Comment by per-samuelsson** [comment]
And for the record, this is nothing that has changed, thats the way they've been implemented always. That makes it possible for the hooking part to abort the transaction, if some condition(s) are not met.

**Comment by careri** [comment]
Ok, what's the time plan for the Transaction Log events?
